### DISCUSSION TITLE
Python 3.10 installed, KnobKraft-orm can't find it

### DISCUSSION THREAD

**Discussion by rssalerno** [christofmuc/KnobKraft-orm#154] [Q&A] [answered]
Where does KnobKraft-orm expect to find Python 3.10?  I have KnobKraft-orm 1.14.1 and Python 3.10.2 and installed on my Mac (OSX High Sierra 10.13.6).  Also, I'm not sure I installed KnobKraft-orm correctly, I dragged the .app and the bin and python folders into the Applications folder.  Is this right?  Sorry for the noob questions.  My goal is to create implementations for the Prophet 5 rev 3.3, Matrix-12 and Memorymoog LAMM.  Thanks.

**Comment by christofmuc** [comment]
Thank you for asking the question, I had planned to document that somewhat better anyway!

The problem with macOS is that there are so many different ways to install Python, and I didn't find out how to bundle it correctly as a Framework.

The following ways to install python I know:

  * There is a Python already shipped with macOS. That is somewhere in the Frameworks folder and usually the wrong version
  * Many people use homebrew and install Python using `brew install python`, but which version you get is a bit random, as they tend to ship latest and we really need the exact version (major.minor) we compiled the Orm for.
  * The most controlled way is to head over to python.org and download an installer for the correct version: https://www.python.org/downloads/macos/

The code in the Orm is found in the directory `adaptions` in the file GenericAdaptation.cpp, lines 214ff , the #ifdef __APPLE__ block.

  1. The first path that is checked is: `/Library/Frameworks/Python.framework/Versions/3.10`.  This is the location the Python installer downloaded from python.org should install it to.
  2. If that is not found, the code checks `/usr/local/opt/python3/Frameworks/Python.framework/Versions/3.10`, which supposedly is where the homebrew installer installs it when running `brew install python`.

More in-depth fun, more to remind myself:

  * https://faun.pub/the-right-way-to-set-up-python-on-your-mac-e923ffe8cf8e
  * https://opensource.com/article/19/5/python-3-default-mac
  * https://docs.python-guide.org/starting/install3/osx/

I would recommend using the installer downloaded from python.org, that should give you the best control.

And yes, the KnobKraft Orm can be installed by dragging it into the application folder, I might even remember the python folders and the bin folder is a bug and shouldn't be there.

**Reply by markusschloesser** [reply]
> * There is a Python already shipped with macOS. That is somewhere in the Frameworks folder and usually the wrong version

AFAIK, not any more. Iirc they stopped that a year ago

**Reply by christofmuc** [reply]
Interesting! The question was about High Sierra, though, there might still be a Python around.

**Reply by markusschloesser** [reply]
https://www.macrumors.com/2022/01/28/apple-removing-python-2-in-macos-12-3/

**Comment by rssalerno** [comment]
My mac actually has Python 2.7.16 installed (which I assume was the OSX default), and also had 3.3 installed from another project which I now updated to 3.10.2.  I will keep trying & report what if anything works.  Thanks.

**Reply by rssalerno** [reply]
I'm not having much luck unfortunately.  I removed the brew versions of python and installed 3.10.0 from python.org.  Now KnobKraft crashes when I double-click on it.  I pasted part of the crash report below.  I'm running OSX 10.13.6 yet /Applications/KnobKraftOrm.app/Contents/MacOS/../Frameworks/libpython3.10.dylib was built for Mac OS X 10.15.  Yet prior to this, KnobKraft was running with native synths but giving an error that it couldn't find python.  Now it doesn't run at all.  How can I troubleshoot this?

Process:               KnobKraftOrm [16856]
Path:                  /Applications/KnobKraftOrm.app/Contents/MacOS/KnobKraftOrm
Identifier:            ???
Version:               ??? (1.15.0)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           KnobKraftOrm [16856]
User ID:               501

Date/Time:             2022-05-09 14:39:48.469 -0400
OS Version:            **Mac OS X 10.13.6** (17G11023)
Report Version:        12
Anonymous UUID:        8CFF233E-439D-2732-2242-2B70A56C8F86


Time Awake Since Boot: 2100 seconds

System Integrity Protection: enabled

Crashed Thread:        0  JUCE Message Thread  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    DYLD, [0x4] Symbol missing

Dyld Error Message:
  Symbol not found: ____chkstk_darwin
  Referenced from: /Applications/KnobKraftOrm.app/Contents/MacOS/../Frameworks/libpython3.10.dylib **(which was built for Mac OS X 10.15)**
  Expected in: /usr/lib/libSystem.B.dylib

**Comment by rssalerno** [comment, marked as answer]
Ok it seems KnobKraft 1.5.0 and 1.14.1 do not run on OSX 10.13.6.

Version 1.13.4 does, with python@3.8 installed as follows:

`$ brew install python@3.8`
`$ cd /usr/local/opt/`
`$ ln -s python@3.8 python3`

**Comment by christofmuc** [comment]
@rssalerno Thank you for the heads up! So this is simlar to the situation on Windows 7, which Python stopped supporting and Windows 7 users are stuck on Python 3.8 unless they try to compile Python themselves for the old OS.

The correct solution would be to bundle the Python as a framework with the app, but my tries so far haven't been successful. Then again, when I target the old mac OS versions, I need to stay on python 3.8 as well, which no longer gets binary releases for the security patches. Sigh. 

Thanks again for your tests!

**Comment by runeflobakk** [comment]
It is a nice feature that the application itself searches some known places where Python may be installed on MacOS 👍 

How about having this as a _configurable_ location if the application is not able to itself locate Python? This would enable people to freely upgrade KnobKraft and work around any new non-standard installation locations. New "known locations" for Python can be added to the application source retroactively. The application could first see if any of the known paths contains Python, and if it can't be located, it could ask the user to browse to locate the directory where Python is installed. It could also offer to browse to python.org to download the installer.

Homebrew currently installs Python 3.10 to `/usr/local/opt/python@3.10/`:
```sh
$ ls -l /usr/local/opt/python@3.10/
total 64
drwxr-xr-x   3 rune  admin     96 Jun  6 14:05 Frameworks
drwxr-xr-x   3 rune  admin     96 Jun  6 14:05 IDLE 3.app
-rw-r--r--   1 rune  admin   3973 Jun 28 13:24 INSTALL_RECEIPT.json
-rw-r--r--   1 rune  admin  13931 Jun  6 14:05 LICENSE
drwxr-xr-x   3 rune  admin     96 Jun  6 14:05 Python Launcher 3.app
-rw-r--r--   1 rune  admin  10138 Jun  6 14:05 README.rst
drwxr-xr-x  15 rune  admin    480 Jun 28 13:24 bin
drwxr-xr-x   3 rune  admin     96 Jun  6 14:05 lib
drwxr-xr-x   4 rune  admin    128 Jun  6 14:05 libexec
drwxr-xr-x   3 rune  admin     96 Jun  6 14:05 share
```

```sh
$ brew info python@3.10
python@3.10: stable 3.10.5 (bottled) [keg-only]
Interpreted, interactive, object-oriented programming language
https://www.python.org/
/usr/local/Cellar/python@3.10/3.10.5 (3,167 files, 57MB) *
  Poured from bottle on 2022-06-28 at 13:23:56
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/python@3.10.rb
License: Python-2.0
==> Dependencies
Build: pkg-config ✔
Required: gdbm ✔, mpdecimal ✔, openssl@1.1 ✔, readline ✔, sqlite ✔, xz ✔
==> Caveats
Python has been installed as
  /usr/local/opt/python@3.10/bin/python3
```

`/usr/local/opt/python@3.10` is a symlink for Python 3.10 maintained by Homebrew, and on my machine currently pointing to my current 3.10 installation at `/usr/local/Cellar/python@3.10/3.10.5`. So `/usr/local/opt/python@3.10` could be a "known path" for KnobKraft to search for Python 3.10 on MacOS.

**Reply by christofmuc** [reply]
Thank you! You're so right in that there should be a last resort setting somewhere in the software in case the detection fails... I have a last resort, but that is an environment variable to set to turn Python off completely, but this is back from the days when the Python integration was an experimental feature - by now, it is doing the main work load of decoding and encoding sysex. 

So yes, I agree we need a manual override settings somewhere to make it find the Python on the Mac, or bundle Python in the DMG so it is self-sustained!

**Reply by conversy** [reply]
I opened an issue #182, I did not see the conversation here.

What if the code looks for an env variable, say  `PYTHONPROG`, and use it as an argument for [`Py_SetProgramName()`](https://docs.python.org/3/c-api/init.html#c.Py_SetProgramName) before calling [`Py_Initialize()`](https://docs.python.org/3/c-api/init.html#c.Py_Initialize)?
`PYTHONPROG` could be set in `Contents/info.plist` in the KnobKraft app bundle:
```
<dict>
    <key>LSEnvironment</key>
    <dict>
        <key>PYTHONPROG</key>
        <string>/whatever/bin/python</string>
    </dict>
```

[Edit] completed the proposed procedure

**Comment by christofmuc** [comment]
Closed - from 2.5.0 on, the DMG contains the correct Python version as a framework.
